### ISSUE TITLE
AutoMoqData runs several times for one test

### ISSUE THREAD

**Issue by Dev-I-Ant** [AutoFixture/AutoFixture#831]
Pasting from [this SO post](https://stackoverflow.com/q/46141256/1679537) for reference :

I've been using AutoFixture / AutoMoqData / Nunit for several months now, quite happily so far.

But I'm getting stuck on something weird:

If I want to run a single test in a test suite (using TestDriven), it seems that the AutoMoqData attribute is launched several times (between 5 and 9, but I don't really know why, and what decides how many times it runs).

The problem is that if I try to freeze an object instance in AutoFixture customizations, it'll end up freezing 5 to 9 objects, using this many fixture instances, and will have no clue which one is injected in my test.

Is there a way to avoid this ? Why does it happen in the first place ?

Sample code :

```csharp
public class AutoMoqDataAttribute : AutoDataAttribute
{
    public AutoMoqDataAttribute()
        : base(new Fixture()
            .Customize(new AutoMoqCustomization())
            .Customize(new FrozenFooCustomization()) // freezing an object
            .Customize(new FrozenFixtureCustomization()) // freezing the passed fixture to have the fixture instance available in the test body
            .Customize(new StaticBindingCustomization()) // binding the frozen object to a static reference
            )
    {
    }
}

public class FrozenFooCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Freeze<Foo>();
    }
}

public class Foo { }

public class FrozenFixtureCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Inject(fixture);
    }
}

public class StaticBindingCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Freeze<Mock<ISomeUtilities>>();
        var someUtilities = fixture.Create<Mock<ISomeUtilities>>();
        someUtilities.Setup(x => x.GetStuff()).Returns(fixture.Create<Foo>());
        // this is executed several times, each time with a different instance of fixture, returning a different instance of Foo
        SomeExtensions.Factory = x => someUtilities.Object;
        // so the last "someUtilities.Object" is bound to the factory, but the test relies on a frozen version generated by one of the previous fixture instances
    }
}

[TestFixture]
public class Test
{
    [Test, AutoMoqData]
    public void SomeTest(
        [Frozen]Mock<ISomeUtilities> objMock,
        SomeObjectToTest sut)
    {
        Assert.IsTrue(ReferenceEquals(objMock.Object.GetStuff(), sut.GetStuff()));
    }
}

public class SomeObjectToTest { }

public static class SomeExtensions
{
    public static Func<object, ISomeUtilities> Factory { get; set; } = x => new SomeUtilities(x);

    public static object GetStuff(this object obj)
    {
        return Factory(obj).GetStuff();
    }
}

public class SomeUtilities : ISomeUtilities
{
    private readonly object _obj;

    public SomeUtilities(object obj)
    {
        _obj = obj;
    }

    public object GetStuff()
    {
        return _obj; // whatever
    }
}

public interface ISomeUtilities
{
    object GetStuff();
}

```

EDIT: SomeExtensions.Factory is a factory used to make extensions mockable, based on [Daniel Cazzulino's approach](http://blogs.clariusconsulting.net/kzu/how-to-mock-extension-methods/).

I'm not a big fan of those extensions and hope to refactor them at some point, but this is for another time. The question here is why are there several fixture instances, making the frozen SomeUtilities mock used in the factory different than the one used in the test, and thus the test to fail.

According to @ploeh, this might come from a defect in either Nunit3 or AutoFixture.Nunit3.

**Comment by zvirja** [comment]
@Dev-I-Ant Thank you for registering this issue!

As far as I got it, your scenario can be simplified to the one below. Here I also capture the value to a static variable during the fixture construction:
```csharp
public class CustomDataAttribute : AutoDataAttribute
{
    public static IFixture LastCreatedFixture;

    public CustomDataAttribute():this(new Fixture())
    {
    }

    private CustomDataAttribute(IFixture fixture):base(fixture)
    {
        LastCreatedFixture = fixture;
    }
}

[TestFixture]
public class Test
{
    [Test, CustomData]
    public void SomeTest(IFixture fixture)
    {
        Assert.AreEqual(fixture, CustomDataAttribute.LastCreatedFixture);
    }
}
```

This test also fails on NUnit3 and passes on xUnit2. 

The issue happens because our custom Data attributes might be created for a multiple times during the test execution and that purely depends on the test framework implementation. For instance when I used xUnit and R# runner, the attribute has been created for 2 times. For NUnit3 + R# I got 3 creations, but I believe that it depends.

I'd like to divide my reply into 2 parts.

## The existing behavior is not a bug
We cannot somehow guarantee that our custom attribute will be created only once as we simply cannot control that. I believe that the test engine will not agree to commit under that neither as a lot of different stuff is going on under the hood - discovery, execution, runner notification, etc.

Another story is concurrency. Engine could use different strategies to run multiple tests even if you disabled the concurrency. One strategy might be e.g. to resolve all the argument values ahead of time and later execute tests one by one, while other is to resolve values immediately before the test execution. From the usual test perspective it usually doesn't matter as the sequential order is guaranteed, while the current scenario will be affected.

The reason why it works for xUnit is that engine uses the values from the latest constructed argument to run tests, while for nUnit that isn't so and extra-instance is created for internal needs. As result, for xUnit the test passes. However, that is a pure side effect usage and you should never rely on that in your tests as that could vary among different runner versions.

If you indeed want to write a reliable test, you should perform the statics setup in the method body instead (disabling the concurrency before):
```csharp
[Test, CustomData]
public void SomeTest(IFixture fixture)
{
    SetupStaticExtensionFactory(fixture);
    
    // rest ...
}
```

Now you don't care when and how many times the fixture was created - you always use the same instance as other parts of the test.

## Room for improvements

Actually, your finding spotted something I never thought before - attribute might be constructed for a multiple times, while actual argument value resolve could happen not for each time. As result, the fixture instance we create in the constructor might be sometimes not used. I [recently profiled](818) that area and found that fixture creation and subsequent customization are a bit expensive (while not critically). Inspired by that I created my own `AutoData` attribute with deferred fixture creation and tested it on my real solution. The result impressed me - NCrunch tests discovery time decreased from ~4.5 sec to ~2.2 sec. It should be noticed though that attribute started to look uglier (as always when we optimize something) 😉

We could investigate that area further and probably improve the AutoData attributes to create fixture lazily, only if actual data discovery happens. That will not help with your scenario somehow (while it could as now side effects might look differently), but at least the test discovery will happen a bit quicker.

---

To summarize. I'd consider the existing behavior as not-a-bug and don't see any issues with that. Rather, we should alter the tests to not depend on internal test engine behavior. We could also register a new issue to investigate laziness, but that's a completely different story.

Also I'd like to invite @adamchester @moodmosaic and @ecampidoglio to this discussion so they could share their vision.

@Dev-I-Ant Let me know if I overlooked something or I you don't agree with my judgement 😃

P.S. `FrozenFixtureCustomization` is redundant as AutoFixture [already does](https://github.com/AutoFixture/AutoFixture/blob/3807550e42db12a7ed3aeec5cb40c7e21e0f5533/Src/AutoFixture/Fixture.cs#L72) that internally.

**Comment by Dev-I-Ant** [comment]
@zvirja Thanks for the extensive answer, seems you got it right :)

I'll test all that right away. Also, this fixture lazy loading seems promising :)

**Comment by zvirja** [comment]
@Dev-I-Ant Any updates on this one? Have you tried the suggested approach and have it worked for you?

**Comment by zvirja** [comment]
@Dev-I-Ant Thank you for the follow up :+1: In this case I'm closing this one. I've already created a separate issue to discuss the mentioned optimization.
